### PR TITLE
Fixed occasional errors with clusteraffinities

### PR DESCRIPTION
**Pull request by XiShanYongYe-Chang**
**What type of PR is this?**

/kind flake

**What this PR does / why we need it**:

In the `karmada-scheduler`, the patch is used to update the rb status. Before the update, deep copy is performed on the rb. As a result, the subsequent rb objects are still the old values, and patch errors may occur. 

So I modified this to keep the rb object up-to-date after the update to avoid affecting subsequent patch operations.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by RainbowMango** [comment]
Does this PR try to address the flaking test?

https://github.com/karmada-io/karmada/actions/runs/5449884625/jobs/9914700141

```
Summarizing 2 Failures:
  [INTERRUPTED] [BasicPropagation] propagation testing Job propagation testing [DeferCleanup (Each)] job propagation testing
  /home/runner/work/karmada/karmada/test/e2e/propagationpolicy_test.go:382
  [FAIL] [ClusterAffinities] propagation testing when [ClusterPropagationPolicy] replica scheduling type is Duplicated schedule with multi clusterAffinity [It] propagate clusterRole and then update the cluster label
  /home/runner/work/karmada/karmada/test/e2e/framework/rbac.go:158

Ran 79 of 139 Specs in 578.314 seconds
FAIL! - Interrupted by Other Ginkgo Process -- 77 Passed | 2 Failed | 0 Pending | 60 Skipped
```

**Comment by chaunceyjiang** [comment]
> Does this PR try to address the flaking test?

> https://github.com/karmada-io/karmada/actions/runs/5449884625/jobs/9914700141


Yes.

**Comment by chaunceyjiang** [comment]
Have you already found the root cause of this issue? /cc @XiShanYongYe-Chang

**Comment by XiShanYongYe-Chang** [comment]
> Have you already found the root cause of this issue?

I caught an error log earlier, but I seem to have lost it. When checking logs, it is found that the scheduler updates the status of the rb to an incorrect old status during subsequent update. As a result, the eventhandler does not meet the conditions during check.

**Comment by RainbowMango** [comment]
another one:
https://github.com/karmada-io/karmada/actions/runs/5507318884/jobs/10037376281

File: pkg/scheduler/scheduler.go
Comment (chaunceyjiang) [inline pkg/scheduler/scheduler.go:1]: Can you remind me where the core modifications of this PR are? I feel like there haven't been any substantial changes compared to before.

Comment (XiShanYongYe-Chang) [reply pkg/scheduler/scheduler.go:1]: After the modification, for each patch operation of the rb object in the scheduler, the updated rb object will be in its latest state. This ensures that there will be no old state interference during the next patch.

**Comment by RainbowMango** [comment]
another failing case on master:
https://github.com/karmada-io/karmada/actions/runs/5610539390/job/15200537208

**Comment by XiShanYongYe-Chang** [comment]
This PR is ready, can you help review it?
/cc @RainbowMango @chaunceyjiang

**Review by RainbowMango — COMMENTED** [review]
/assign

File: pkg/scheduler/scheduler.go
Comment (RainbowMango) [reply pkg/scheduler/scheduler.go:1]: Indeed, it's hard to figure out the core modifications.

File: pkg/scheduler/scheduler.go [hunk 1]
```diff
@@ -809,21 +811,22 @@ func patchClusterBindingStatusWithAffinityName(karmadaClient karmadaclientset.In
...
 	if err != nil {
 		return err
 	}
 	if len(patchBytes) == 0 {
 		return nil
 	}
 
-	_, err = karmadaClient.WorkV1alpha2().ClusterResourceBindings().Patch(context.TODO(), crb.Name, types.MergePatchType, patchBytes, metav1.PatchOptions{}, "status")
+	_, err = karmadaClient.WorkV1alpha2().ClusterResourceBindings().Patch(context.TODO(), crb.Name, types.MergePatchType,
+		patchBytes, metav1.PatchOptions{}, "status")
```
Comment (RainbowMango) [inline pkg/scheduler/scheduler.go:829]: I'm afraid it doesn't make readability better.

File: pkg/scheduler/scheduler.go [hunk 2]
```diff
@@ -809,21 +811,22 @@ func patchClusterBindingStatusWithAffinityName(karmadaClient karmadaclientset.In
...
-	updateCRB.Status.SchedulerObservedAffinityName = affinityName
-	return patchClusterResourceBindingStatus(karmadaClient, crb, updateCRB)
+	originalCRB := crb.DeepCopy()
+	crb.Status.SchedulerObservedAffinityName = affinityName
+	return patchClusterResourceBindingStatus(karmadaClient, originalCRB, crb)
 }
 
-func patchClusterResourceBindingStatus(karmadaClient karmadaclientset.Interface, crb, updateCRB *workv1alpha2.ClusterResourceBinding) error {
-	patchBytes, err := helper.GenMergePatch(crb, updateCRB)
+func patchClusterResourceBindingStatus(karmadaClient karmadaclientset.Interface, originalCRB, crb *workv1alpha2.ClusterResourceBinding) error {
```
Comment (RainbowMango) [inline pkg/scheduler/scheduler.go:819]: How about naming the parameter with `originalxxx` and `modifiedxxx`？

File: pkg/scheduler/scheduler.go [hunk 3]
```diff
@@ -809,21 +811,22 @@ func patchClusterBindingStatusWithAffinityName(karmadaClient karmadaclientset.In
 	}
 
 	klog.V(4).Infof("Begin to patch status with affinityName(%s) to ClusterResourceBinding(%s).", affinityName, crb.Name)
-	updateCRB := crb.DeepCopy()
-	updateCRB.Status.SchedulerObservedAffinityName = affinityName
-	return patchClusterResourceBindingStatus(karmadaClient, crb, updateCRB)
+	originalCRB := crb.DeepCopy()
+	crb.Status.SchedulerObservedAffinityName = affinityName
+	return patchClusterResourceBindingStatus(karmadaClient, originalCRB, crb)
```
[the remark this answers is not in the document]
Comment (XiShanYongYe-Chang) [reply pkg/scheduler/scheduler.go]: updated

**Comment by XiShanYongYe-Chang** [comment]
/hold
Let me think about whether it is possible to consolidate the update operations as much as possible, in order to reduce conflicts and impacts caused by multiple updates in different places.

**Comment by XiShanYongYe-Chang** [comment]
I think using the patch method in #4094 should be able to solve the current issue with patch errors. What do you think? @chaunceyjiang

**Comment by XiShanYongYe-Chang** [comment]
It seems that the issue has not occurs for a long time.
/close

**Pull request closed without merging**